File: extensions/shared/library/src/main/java/app/revanced/extension/shared/patches/SanitizeSharingLinksPatch.java
package app.revanced.extension.shared.patches;

import app.revanced.extension.shared.settings.BaseSettings;

/**
 * YouTube and YouTube Music.
 */
@SuppressWarnings("unused")
public final class SanitizeSharingLinksPatch {
    private static final String NEW_TRACKING_PARAMETER_REGEX = ".si=.+";
    private static final String OLD_TRACKING_PARAMETER_REGEX = ".feature=.+";

    /**
     * Injection point.
     */
    public static String sanitize(String url) {
        if (BaseSettings.SANITIZE_SHARED_LINKS.get()) {
            url = url
                    .replaceAll(NEW_TRACKING_PARAMETER_REGEX, "")
                    .replaceAll(OLD_TRACKING_PARAMETER_REGEX, "");
        }

        if (BaseSettings.REPLACE_MUSIC_LINKS_WITH_YOUTUBE.get()) {
            url = url.replace("music.youtube.com", "youtube.com");
        }

        return url;
    }
}

File: patches/src/main/kotlin/app/revanced/patches/music/misc/privacy/SanitizeSharingLinksPatch.kt
package app.revanced.patches.music.misc.privacy

import app.revanced.patches.music.misc.extension.sharedExtensionPatch
import app.revanced.patches.music.misc.settings.PreferenceScreen
import app.revanced.patches.music.misc.settings.settingsPatch
import app.revanced.patches.shared.misc.privacy.sanitizeSharingLinksPatch

@Suppress("unused")
val sanitizeSharingLinksPatch = sanitizeSharingLinksPatch(
    block = {
        dependsOn(
            sharedExtensionPatch,
            settingsPatch,
        )

        compatibleWith(
            "com.google.android.apps.youtube.music"(
                "7.29.52",
                "8.10.52"
            )
        )
    },
    preferenceScreen = PreferenceScreen.MISC,
    replaceMusicLinksWithYouTube = true
)